src/IO.php

```diff
@@ -0,0 +1,170 @@
+        $this->unsafe = $f;
+    }
+
+    /**
+     * @template B
+     * @param B $x
+     * @return IO<B>
+     */
+    public static function of($x): IO
+    {
+        return new self(static fn () => $x);
+    }
+
+    /**
+     * @template B
+     * @param HK1<IOBrand, B> $b
+     * @return IO<B>
+     *
+     * @psalm-pure
+     */
+    public static function fromBrand(HK1 $b): IO
+    {
+        /** @var IO $b */
+        return $b;
+    }
+
+    /**
+     * @template B
+     * @param Monad<IOBrand> $monad
+     * @param callable(A): HK1<IOBrand, B> $f
+     * @return IO<B>
+     */
+    public function ibind(Monad $monad, callable $f): IO
+    {
+        /** @psalm-suppress ArgumentTypeCoercion */
+        return self::fromBrand($monad->bind($this, $f));
+    }
+
+    /**
+     * @template B
+     * @param callable(A): HK1<IOBrand, B> $f
+     * @return IO<B>
+     *
+     * @psalm-suppress LessSpecificImplementedReturnType
+     */
+    public function bind(callable $f): IO
+    {
+        return $this->ibind(new IOMonad(), $f);
+    }
+
+    /**
+     * @template B
+     * @param Applicative<IOBrand> $applicative
+     * @param B $a
+     * @return IO<B>
+     */
+    public static function ipure(Applicative $applicative, $a): IO
+    {
+        return self::fromBrand($applicative->pure($a));
+    }
+
+    /**
+     * @template B
+     * @param B $a
+     * @return IO<B>
+     *
+     * @psalm-suppress LessSpecificImplementedReturnType
+     */
+    public static function pure($a): self
+    {
+        return self::ipure(new IOApplicative(), $a);
+    }
+
+        /**
+     * @template B
+     * @param Apply<IOBrand> $apply
+     * @param HK1<IOBrand, callable(A): B> $f
+     * @return IO<B>
+     *
+     * @psalm-mutation-free
+     */
+    public function iapply(Apply $apply, HK1 $f): IO
+    {
+        /** @psalm-suppress ArgumentTypeCoercion */
+        return self::fromBrand($apply->apply($f, $this));
+    }
+
+    /**
+     * @template B
+     * @param HK1<IOBrand, callable(A): B> $f
+     * @return IO<B>
+     *
+     * @psalm-mutation-free
+     *
+     * @psalm-suppress LessSpecificImplementedReturnType
+     */
+    public function apply(HK1 $f): IO
+    {
+        return $this->iapply(new IOApply(), $f);
+    }
+
+    /**
+     * @template B
+     * @param Functor<IOBrand> $functor
+     * @param callable(A): B $f
+     * @return IO<B>
+     *
+     * @psalm-mutation-free
+     */
+    public function imap(Functor $functor, callable $f): IO
+    {
+        /** @psalm-suppress ArgumentTypeCoercion */
+        return self::fromBrand($functor->map($f, $this));
+    }
+
+    /**
+     * @template B
+     * @param callable(A): B $f
+     * @return IO<B>
+     *
+     * @psalm-mutation-free
+     *
+     * @psalm-suppress LessSpecificImplementedReturnType
+     */
+    public function map(callable $f): IO
+    {
+        return $this->imap(new IOFunctor(), $f);
+    }
+
+    /**
+     * @todo Should we have a parameter here?
```

Comment:
no, no parameters are needed here, this amounts to invoking a function with no arguments

Reply:
Done.